### DISCUSSION TITLE
Tor authentication

### DISCUSSION THREAD

**Discussion by DaveyHeisen** [JarryShaw/darc#51] [Q&A] [answered]
Hi,

Thanks for getting back to me, I have now managed to get the application running using the docker-compose.debug docker files and starting the container with the CLI parameter.

However, I have read the documentation and cannot find anything about the TOR authentication it is asking for.
Is there any documentation available, or can you help me on this one?

Thanks!

**Comment by JarryShaw** [comment, marked as answer]
The `TOR_PASS` environment variable is for configuring the Tor authentication required by Tor Controller instance, which is used to manipulate Tor configuration and behaviour at runtime dynamically in a programming way.

Currently, `darc` uses Tor Controller to renew its Tor relay at the end of each processing *round*, as such to avoid back tracing through network activity.

For more information about Tor Controller, please refer to documentations of Tor and `stem`, the Python library for managing Tor applications.

**Comment by DaveyHeisen** [comment]
Okay, I have build the Docker image using the docker-compose.debug files, I then run the image using the following command: 
docker run -i jsnbzh/darc

It then requires me to fill in the Tor Authentication, when I do so, the container shows the darc --help text and the container closes.

Can you please tell what command I have to run to make it crawl the data from a text file and save the data to SQL database?

**Reply by JarryShaw** [reply]
Do please refer to the documentation of `darc` project itself. You can specify the credential for Tor Controller through the `TOR_PASS` environment variable; it’s value is set in the `.torrc` configurations.

**Comment by DaveyHeisen** [comment]
Hi JarryShaw,

Thanks for your help so far. Any clue why the connection is refused on TOR_CTRL port 9051?

crawler    | /app/darc/proxy/tor.py:94: TorRenewFailed: [Errno 111] Connection refused
crawler    |   _TOR_CTRL = stem.control.Controller.from_port(port=int(TOR_CTRL))

Thanks again!

**Reply by JarryShaw** [reply]
Check if the port is open; if your Tor Controller daemon is up and running; if you are using the right authentication credentials. And if all checked, consult SO or the Tor community instead. Bedankt!

**Comment by vpnmastersecure** [comment]
Thanks, very useful!
